### PR TITLE
DBZ-6178 Connector changes for Spanner Postgres Dialect 

### PR DESCRIPTION
**Pull request by imtj1**
Contains changes for supporting Postgres Dialect Spanner in connector.

Bug ID : https://issues.redhat.com/browse/DBZ-6178

### CODE REVIEW

**Comment by jpechane** [comment]
@imtj1 Thanks fro the PR! Would it be possible to extract the dialect specific code into an interface and two implementation classes? Then the proper class would be s=instantieated and used in the codebase instead of having ifs and switches dispersed around.
We do similar thing for examplein Oracle - for Xstream and LogMiner implementations. It is different in functionality but sam in principle.

**Comment by imtj1** [comment]
> @imtj1 Thanks fro the PR! Would it be possible to extract the dialect specific code into an interface and two implementation classes? Then the proper class would be s=instantieated and used in the codebase instead of having ifs and switches dispersed around. We do similar thing for examplein Oracle - for Xstream and LogMiner implementations. It is different in functionality but sam in principle.

These are great observations Jiri, thanks for pointing them out.

Although I thought about this and decided it would not add much value due to the following reasons:

1. The primary reason for a branch in code for the Postgres dialect is because the executed query is different (String resource), thereby splitting into interface would mean the implementations would only be different in the query they execute and nothing more. All API calls and logic would remain the same.
2. Spanner does not have plans to release more dialects anytime in the near future, so an interface based on dialect changes is not of much added value at the moment.

Again thanks for your inputs. Appreciate it. :)

**Comment by jpechane** [comment]
@imtj1 hanks for the explanation. Still if pssible I'd prefer the refactoring. The reason is that it is aways conveninet to have those differences centralized. Even if no other dialect is added it is still possible that in future additional differences migh be identified in the current ones that would require additional changes in the code.

File: src/main/java/io/debezium/connector/spanner/db/model/ModType.java [hunk 1]
```diff
@@ -14,5 +14,6 @@
 public enum ModType {
     INSERT,
     UPDATE,
-    DELETE
+    DELETE,
+    UNKNOWN
```
[the remark this answers is not in the document]
Comment (imtj1) [reply src/main/java/io/debezium/connector/spanner/db/model/ModType.java:18]: This is used when there is an unrecognised ModType in Mapper.

**Comment by jpechane** [comment]
@imtj1 Could you please rebase the PR against the latest `main`?

**Comment by nancyxu123** [comment]
Let's hold back on pushing this PR until some performance testing is done.

**Comment by imtj1** [comment]
> 

Done

**Comment by imtj1** [comment]
> @imtj1 Could you please rebase the PR against the latest `main`?

@jpechane Done.

File: src/main/java/io/debezium/connector/spanner/db/mapper/ChangeStreamRecordMapper.java [hunk 2]
```diff
@@ -134,12 +257,145 @@ DataChangeEvent toDataChangeEvent(Partition partition, Struct row,
                 streamEventMetadataFrom(partition, commitTimestamp, resultSetMetadata));
```
Comment (nancyxu123) [inline src/main/java/io/debezium/connector/spanner/db/mapper/ChangeStreamRecordMapper.java:257]: Do we want to use modTypeFrom and valueCaptureTypeFrom here?

File: src/main/java/io/debezium/connector/spanner/db/model/schema/Column.java [hunk 3]
```diff
@@ -44,46 +47,103 @@ public boolean isNullable() {
...
+                throw new IllegalArgumentException("Unknown spanner type " + spannerType);
+            case POSTGRESQL:
+                if (spannerType.endsWith("[]")) {
+                    // Substring "xxx[]"
+                    // Must check array type first
+                    String spannerArrayType = spannerType.substring(0, spannerType.length() - 2);
+                    ColumnType itemType = parseColumnType(spannerArrayType, dialect);
+                    return new ColumnType(DataType.ARRAY, itemType);
+                }
+                if ("BOOLEAN".equals(spannerType)) {
```
Comment (nancyxu123) [inline src/main/java/io/debezium/connector/spanner/db/model/schema/Column.java:102]: I don't see the JsonB data type or the spanner commit timestamp data type.

In addition, do we want to remove the Spanner type length limits?

Comment (imtj1) [reply src/main/java/io/debezium/connector/spanner/db/model/schema/Column.java:102]: Added SPANNER.COMMIT_TIMESTAMP, Referred this file for relevant changes http://shortn/_JT4iJYizUg, but did not have JsonB in it. Spanner type length limits only apply for arrays and they were already in place before and also exist in the referred file above.

File: src/main/java/io/debezium/connector/spanner/db/mapper/ChangeStreamRecordMapper.java (hunk 2 above)
Comment (imtj1) [reply src/main/java/io/debezium/connector/spanner/db/mapper/ChangeStreamRecordMapper.java:257]: Done

**Comment by jpechane** [comment]
@imtj1 Applied, thanks

File: src/main/java/io/debezium/connector/spanner/config/validation/ChangeStreamValidator.java [hunk 4]
```diff
@@ -88,12 +89,23 @@ public ChangeStreamValidator validate() {
...
-        Statement statement = Statement.newBuilder("select change_stream_name " +
-                "from information_schema.change_streams cs " +
-                "where cs.change_stream_name = @streamname")
-                .bind("streamName")
-                .to(streamName)
-                .build();
+        Statement statement;
+        if (databaseClient.getDialect() == Dialect.POSTGRESQL) {
+            statement = Statement.newBuilder("select change_stream_name " +
+                    "from information_schema.change_streams cs " +
```
[the remark this answers is not in the document]
Comment (imtj1) [reply src/main/java/io/debezium/connector/spanner/config/validation/ChangeStreamValidator.java:95]: IMO, it should not be done as part of connector code, it should be taken care of by the connector config ?

**Pull request merged**